packages/lib/services/e2ee/crypto.ts

```diff
@@ -0,0 +1,132 @@
+import { _ } from '../../locale';
+import { Crypto, CryptoBuffer, Digest, CipherAlgorithm, EncryptionResult } from './types';
+import { promisify } from 'util';
+import {
+	getCiphers as nodeGetCiphers, getHashes as nodeGetHashes,
+	randomBytes as nodeRandomBytes,
+	pbkdf2 as nodePbkdf2,
+	createCipheriv, createDecipheriv,
+	CipherGCMOptions, CipherGCM, DecipherGCM,
+} from 'crypto';
+
+const crypto: Crypto = {
+
+	getCiphers: () => {
+		return nodeGetCiphers();
+	},
+
+	getHashes: () => {
+		return nodeGetHashes();
+	},
+
+	randomBytes: async (size: number) => {
+		const randomBytesAsync = promisify(nodeRandomBytes);
+		return randomBytesAsync(size);
+	},
+
+	pbkdf2Raw: async (password: string, salt: CryptoBuffer, iterations: number, keylen: number, digest: Digest) => {
+		const pbkdf2Async = promisify(nodePbkdf2);
+		return pbkdf2Async(password, salt, iterations, keylen, digest);
+	},
+
+
+	encryptRaw: (data: CryptoBuffer, algorithm: CipherAlgorithm, key: CryptoBuffer, iv: CryptoBuffer | null, authTagLength: number, associatedData: CryptoBuffer | null) => {
```

Comment:
Make all parameters required - https://joplinapp.org/help/dev/coding_style#avoid-default-and-optional-parameters

It seems you're always setting the iv in the calling code anyway?

And what is associatedData? Unless I'm missing something it looks like you always set it to null so why is it even needed as a parameter?

Reply:
> Make all parameters required

I've added a new interface `EncryptionParameters` which contains all parameters, and all the parameters are explicitly set in `encrypt()`/`decrypt()` of `EncryptionService`.

> It seems you're always setting the iv in the calling code anyway?

Yes.

> And what is associatedData? Unless I'm missing something it looks like you always set it to null so why is it even needed as a parameter?

It is also a crypto parameter. As you required I have added it to the `EncryptionParameters` and set it in `EncryptionService`.